yarn-project/sequencer-client/src/publisher/viem-tx-sender.ts

```diff
@@ -0,0 +1,165 @@
+import { TxSenderConfig } from './config.js';
+import { L1ProcessArgs as ProcessTxArgs, L1PublisherTxSender } from './l1-publisher.js';
+import { CompleteContractData, UnverifiedData } from '@aztec/types';
+import { createDebugLogger } from '@aztec/foundation';
+import {
+  GetContractReturnType,
+  Hex,
+  HttpTransport,
+  PublicClient,
+  WalletClient,
+  createPublicClient,
+  createWalletClient,
+  getAddress,
+  getContract,
+  http,
+} from 'viem';
+import { RollupAbi, UnverifiedDataEmitterAbi } from '@aztec/l1-contracts/viem';
+import { PrivateKeyAccount, privateKeyToAccount } from 'viem/accounts';
+import * as chains from 'viem/chains';
+
+/**
+ * Pushes transactions to the L1 rollup contract using viem.
+ */
+export class ViemTxSender implements L1PublisherTxSender {
+  private rollupContract: GetContractReturnType<typeof RollupAbi, PublicClient, WalletClient>;
+  private unverifiedDataEmitterContract: GetContractReturnType<
+    typeof UnverifiedDataEmitterAbi,
+    PublicClient,
+    WalletClient
+  >;
+  private publicClient: PublicClient<HttpTransport, chains.Chain>;
+
+  private log = createDebugLogger('aztec:sequencer:viem-tx-sender');
+  private account: PrivateKeyAccount;
+  private chain: chains.Chain;
+
+  constructor(config: TxSenderConfig) {
+    const {
+      rpcUrl,
+      chainId,
+      publisherPrivateKey,
+      rollupContract: rollupContractAddress,
+      unverifiedDataEmitterContract: unverifiedDataEmitterContractAddress,
+    } = config;
+
+    this.account = privateKeyToAccount(('0x' + publisherPrivateKey.toString('hex')) as Hex);
```

Comment:
weird that you have to cast this as `Hex`.
even their example shows that it should work https://viem.sh/docs/accounts/privateKey.html#usage

Reply:
This was the issue:
<img width="821" alt="image" src="https://user-images.githubusercontent.com/13470840/234834648-dd00b6b8-3bdb-4eda-abe4-4f65ea3f88c8.png">

and when I use this syntax:

<img width="606" alt="image" src="https://user-images.githubusercontent.com/13470840/234836195-bde9a652-7f53-4aa1-afff-1054d2a48ba9.png">

everything worked.

I am surprised that TS is allowing for this kind of type check.

Reply:
[Template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html), added in [4.1](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/) ;-)